modules/auxiliary/scanner/scada/bacnet_l3.rb

```diff
@@ -0,0 +1,412 @@
+  DISCOVERY_MESSAGE_L2_LEN = Array[DISCOVERY_MESSAGE_L2.length].pack('n')
+
+  READ_MULTIPLE_DEVICES_PROP = "\x1e\x09\x08\x1f".freeze
+  READ_MODEL_NAME_PROP = "\x19\x46".freeze
+  READ_FIRMWARE_VERSION_PROP = "\x19\x2c".freeze
+  READ_APP_SOFT_VERSION_PROP = "\x19\x0c".freeze
+  READ_DESCRIPTION_PROP = "\x19\x1c".freeze
+
+  GET_PROPERTY_MESSAGES_L3_SIMPLE = [
+    "\x81\n\u0000\u0011\u0001\u0004\u0002\u0002\u0000\f\f\u0002{object_identifier}#{READ_MODEL_NAME_PROP}", # model-name
+    "\x81\n\u0000\u0011\u0001\u0004\u0002\u0002\u0000\f\f\u0002{object_identifier}#{READ_FIRMWARE_VERSION_PROP}", # firmware-revision
+    "\x81\n\u0000\u0011\u0001\u0004\u0002\u0002\u0000\f\f\u0002{object_identifier}#{READ_APP_SOFT_VERSION_PROP}", # application-software-version
+    "\x81\n\u0000\u0011\u0001\u0004\u0002\u0002\u0000\f\f\u0002{object_identifier}#{READ_DESCRIPTION_PROP}"
+  ].freeze # description
+
+  GET_PROPERTY_MESSAGES_L3_NESTED = [
+    "\u0001${dest_net_id}{dadr_len}{dadr}\xFF\u0002\u0002\u0002\f\f\u0002{object_identifier}#{READ_MODEL_NAME_PROP}",
+    "\u0001${dest_net_id}{dadr_len}{dadr}\xFF\u0002\u0002\u0002\f\f\u0002{object_identifier}#{READ_FIRMWARE_VERSION_PROP}",
+    "\u0001${dest_net_id}{dadr_len}{dadr}\xFF\u0002\u0002\u0002\f\f\u0002{object_identifier}#{READ_APP_SOFT_VERSION_PROP}",
+    "\u0001${dest_net_id}{dadr_len}{dadr}\xFF\u0002\u0002\u0002\f\f\u0002{object_identifier}#{READ_DESCRIPTION_PROP}"
+  ].freeze
+
+  # Global variables initialization
+  $iface
+  $timeout
+  $packet_count
+  $port
+
+  def initialize
+    super(
+      'Name' => 'BACnet Scanner',
+      'Description' => '
+        Discover BACnet devices by broadcasting Who-is message, then poll
+        discovered devices for properties including model name,
+        software version, firmware revision and description.
+      ',
+      'Author' => ['Paz @ SCADAfence'],
+      'License' => MSF_LICENSE,
+      'Notes' => {
+        'Stability' => [CRASH_SAFE],
+        'Reliability' => [UNRELIABLE_SESSION],
+        'SideEffects' => [SCREEN_EFFECTS]
+      }
+    )
+
+    register_options(
+      [
+        OptInt.new('TIMEOUT', Array[true, 'The socket connect timeout in seconds', DEFAULT_SERVER_TIMEOUT]),
+        OptInt.new('COUNT', Array[true, 'The number of times to send each packet', DEFAULT_SEND_COUNT]),
+        OptPort.new('PORT', Array[true, 'BACnet/IP UDP port to scan (usually between 47808-47817)', DEFAULT_BACNET_PORT]),
+        OptString.new('INTERFACE', Array[true, 'The interface to scan from', 'eth1'])
+      ], self.class
+    )
+  end
+
+  def hex_to_bin(s)
+    s.scan(/../).map { |x| x.hex.chr }.join
+  end
+
+  def bin_to_hex(s)
+    s.each_byte.map { |b| b.to_s(16).rjust(2, '0') }.join
+  end
+
+  def to_ip_address(s)
+    s = s.unpack('C*')
+    "#{s[0]}.#{s[1]}.#{s[2]}.#{s[3]}"
+  end
+
+  # Create UDP packet with the l2 and l3 details.
+  def create_packet_base
+    udp_pkt = PacketFu::UDPPacket.new
```

Comment:
Is using `PacketFu` to forge packets and send broadcast necessary? Maybe it is possible to simply use the [`Rex::Socket::Udp`](https://github.com/rapid7/rex-socket/blob/master/lib/rex/socket/udp.rb) instead?

Reply:
I used `PacketFu` so I could read/write MAC and IP addresses easily, and to align the code with byte representation.